### ISSUE TITLE
Regression/Classification for multiple outputs, but with training data that has incomplete labels

### ISSUE THREAD

**Issue by jmacglashan** [deeplearning4j/deeplearning4j#1380]
Consider a regression problem with multiple output values. Under the typical supervised learning setting, we receive a dataset that has inputs and all target output values for that dataset. For example, for a problem with 2 inputs and 2 outputs, a training example would look like ((x1, x2), (y1, 2)).

DL4J handles learning from this kind of dataset fine. However, a slightly different training setting that DL4J does not seem to support is when each training example only provides the target value for some subset of the output variables. For example, one training example might only provide the y1 label: ((x1, x2), (y1, ?)).

The way this should be handled is that you compute the error gradient with respect to y1 alone (or whatever the subset of provided outputs is) and do gradient descent with that. As long as the set of all training examples gives information about all the output variables, the full multi-variable output function will be learned.

I had hoped that the label masks would provide this functionality, but it does not, and it would be great to have this support.

For what it's worth, one place where this partial output labels would be used is in Deep Q-learning, where you have a different output for each action (representing the Q-value for that action), but each example from the training data only ever provides a target value for one of those actions.

**Comment by AlexDBlack** [comment]
Per-output masking has been supported for a while now - that gives you the type of functionality mentioned here. This issue should have been closed then, but was missed.
https://github.com/deeplearning4j/deeplearning4j/pull/2993
